### ISSUE TITLE
[BUG] 2.3.0-A with --psy-rd >0 produces artifacts and inconsistent results with identical source and settings.

### ISSUE THREAD

**Issue by igorbaryshev** [psy-ex/svt-av1-psy#117]
**Overview**
 - Random block-shaped artifacts at the exact same frames:
![mpv_uPNj9akV41](https://github.com/user-attachments/assets/43619563-9799-4684-b84f-a808c78e23d1)
![mpv_NTlz8ectta](https://github.com/user-attachments/assets/61f69e42-83ba-4348-ad43-4a6beb191fd6)
![mpv_D7bapSefgH](https://github.com/user-attachments/assets/3d166731-30c8-419b-a26c-8e2cde5334c8)
![mpv_KUGXi5Ftk7](https://github.com/user-attachments/assets/f4961ac5-05c7-4909-b4fd-9d9c5d8ad34c)
- A screenshot from another encode to make sure it's not just one video issue:
![image](https://github.com/user-attachments/assets/819dc1fb-8480-4475-bb0e-f04dd513f068)
 
 - File sizes are inconsistent as well, obviously.
![image](https://github.com/user-attachments/assets/0bb474e0-eb8d-4ca2-94b4-b2c656774857)

- Presets 5, 6, 7 are the most affected. The slowest I tried was 3 and it wasn't as noticeable there. Preset 10 I couldn't really tell.
- Starts to become visible at `--psy-rd` values around 0.5 and up, but file size inconsistencies are present with any non-0 value at any preset I tried.
- CRF's tried were 35 and 22.5

**Branch**
In which branch does the issue appear to be occurring?
- [x] `master`
- [ ] `testing`

**Reproduction**
1. Encode with `--psy-rd 6.0` and `--preset 6`
2. Watch the result

**Expected behavior**
No obvious artifacts, consistent encoding results.

**Platform (please complete the following information):**
 - Windows 11 24H2
 - Clang 19.1.6, and another build by @Andarwinux with Clang 20.0.0git
 - Build command for my Clang 19.1.6 build:
 ```powershell
cmake -B svt_build --fresh -T ClangCL -DBUILD_SHARED_LIBS=OFF -DENABLE_AVX512=OFF -DLIBDOVI_FOUND=1 -DLIBHDR10PLUS_RS_FOUND=1 -DCMAKE_CXX_FLAGS_RELEASE="/DNDEBUG /clang:-O2 -flto -march=znver1 -fprofile-use=D:\PGO\IR+CSIR.profdata" -DCMAKE_C_FLAGS_RELEASE="/DNDEBUG /clang:-O2 -flto -march=znver1 -fprofile-use=D:\PGO\IR+CSIR.profdata" && cmake --build svt_build --config Release
```

**Version String (please complete the following information):**
 - Clang 19.6.1 build
```powershell
SVT-AV1-PSY v2.3.0-A-dirty (release)
PSY Release: A
```
 - Clang 20.0.0git build
 ```powershell
SVT-AV1-PSY v2.3.0-A (release)
PSY Release: A
```

**Additional context / Relevant Files**
 - Was only able to test 10-bit encoding, because 8-bit encodes crash at the very beginning after random number of frames, like up to 400, sometimes as low as 44, average ~120.
 - I did several encodes with my Clang 19.1.6 build to verify that it's not just a build difference when it comes to resulting file size and artifact _randomness_.
 - With `--psy-rd 0` files seem to be identical from encode to encode.

**Comment by BlueSwordM** [comment]
Thank you for the bug report. I will try to reproduce it on my end (CachyOS **Linux**).

**Comment by igorbaryshev** [comment]
> Did you try weaker settings (0.5 to 4.0)?

Yep, tried 3.0 and 4.0, same story. Just as I said, anything at 0.5 and up becomes pretty visible on my side.

**Comment by igorbaryshev** [comment]
`--psy-rd 1.0`
![mpv_WrGAL63L71](https://github.com/user-attachments/assets/93edb0b0-bc2c-4a4a-81a6-7c9611606ff4)

**Comment by kuzon3** [comment]
im getting the same bug even at 0.5. 
built my own svt using the build instructions for windows.

**Comment by Ironclad17** [comment]
I saw the same occasionally in very noisy, heavily filtered scenes with --psy-rd 1.0. Lowering to crf 10.0 appeared to resolve it, but it may have been random.

**Comment by gianni-rosato** [comment]
Hi,

Would you be able to provide a couple of things to assist in our diagnosis?

- A sample clip
- Encoding settings to reproduce the issues you describe
- Test conditions that reproduce the bug on Linux and/or macOS

I'm able to reproduce the issue with inconsistent run-to-run behavior on Linux, but I cannot reproduce any visual issues.

We should probably make this more clear in our documentation, but it is worth noting that SVT-AV1-PSY is not officially supported for Windows and there are no plans to change this any time soon. We're most interested in reproducing the visual inconsistencies you present here in Linux/macOS environments. We greatly appreciate your help.

**Comment by igorbaryshev** [comment]
- Sample clip ->[friends.zip](https://github.com/user-attachments/files/18289984/friends.zip)
*.y4m file was over the size limit packed in a 7z inside zip, so I'm providing the clip that I made the .y4m from with this command: `ffmpeg -i friends.mkv -strict -1 friends.y4m`
- The simplest encoding settings successfully reproduce the issue on Ubuntu 24.10:
`SvtAv1EncApp --input friends.y4m --output friends.ivf --preset 6 --psy-rd 6.0`
- Test conditions were just basic Ubuntu 24.10 and the app itself, built with `build.sh -x` from `testing` branch with latest commit being https://github.com/psy-ex/svt-av1-psy/commit/fbe4e20af7b7b3bfa0a957e217feb2409f849fdc (just wanted to see whether it got fixed somehow in `testing` vs `master`)

**Comment by BlueSwordM** [comment]
> * Sample clip ->[friends.zip](https://github.com/user-attachments/files/18289984/friends.zip)
>       *.y4m file was over the size limit packed in a 7z inside zip, so I'm providing the clip that I made the .y4m from with this command: `ffmpeg -i friends.mkv -strict -1 friends.y4m`
> 
>     * The simplest encoding settings successfully reproduce the issue on Ubuntu 24.10:
>       `SvtAv1EncApp --input friends.y4m --output friends.ivf --preset 6 --psy-rd 6.0`
> 
>     * Test conditions were just basic Ubuntu 24.10 and the app itself, built with `build.sh -x` from `testing` branch with latest commit being [fbe4e20](https://github.com/psy-ex/svt-av1-psy/commit/fbe4e20af7b7b3bfa0a957e217feb2409f849fdc) (just wanted to see whether it got fixed somehow in `testing` vs `master`)

Oh thank you, I am now able to reproduce the issue.
On my sources, even at P9, I wasn't able to reproduce this.

Thanks a lot :heart:

**Comment by BlueSwordM** [comment]
I was able to fix build issues, and have found new findings.

With this new patch in the testing branch:
https://github.com/psy-ex/svt-av1-psy/commit/a3f3ca19924d1563488dfae57b53f080c7fab529
 
A bug that a user messaged me personally was fixed by changing the absolute value conversion, which fixed the your bug at Preset 5 and Preset >8.
 
 However, the issue still persists on Preset 6 and Preset 7.
 I can't reproduce the issue with high --psy-rd values, but not low --psy-rd values.

In summary:
1. Fixed some buggy code that was only visible in debug mode with the YASM assembler; NASM couldn't detect the undefined code.
2. The issue with `--psy-rd`can only be replicated with values >0.25.
3. The issue with `--psy-rd` can only be replicated with Preset 6-7.

My best guess? Overflowing values with the RDO calculations for the rest of the encoded block artifacts, causing the encoder to believe they have so much distortion to blur them to pieces.

**Comment by igorbaryshev** [comment]
@BlueSwordM, I'll get to checking it out in a bit, for sure!
Meanwhile, got another one for ya https://github.com/psy-ex/svt-av1-psy/issues/122 😁

**Comment by igorbaryshev** [comment]
@BlueSwordM the issue is still present for me with latest testing commit.
This is on preset 3 with psy-rd 3:
![image](https://github.com/user-attachments/assets/1a802b64-81c0-493c-843e-ba142374f55c)

**Comment by BlueSwordM** [comment]
> I was able to fix build issues, and have found new findings.
> 
> With this new patch in the testing branch: [a3f3ca1](https://github.com/psy-ex/svt-av1-psy/commit/a3f3ca19924d1563488dfae57b53f080c7fab529)
> 
> A bug that a user messaged me personally was fixed by changing the absolute value conversion, which fixed the your bug at Preset 5 and Preset >8.
> 
> However, the issue still persists on Preset 6 and Preset 7. I can't reproduce the issue with high --psy-rd values, but not low --psy-rd values.
> 
> In summary:
> 
>     1. Fixed some buggy code that was only visible in debug mode with the YASM assembler; NASM couldn't detect the undefined code.
> 
>     2. The issue with `--psy-rd`can only be replicated with values >0.25.
> 
>     3. The issue with `--psy-rd` can only be replicated with Preset 6-7.
> 
> 
> My best guess? Overflowing values with the RDO calculations for the rest of the encoded block artifacts, causing the encoder to believe they have so much distortion to blur them to pieces.

Better news: my guess has become reality.
I found the issue: overflows were happening since the variable types we picked were... inadequate for some cases in LBD (8-bit) and very inadequate in HBD (10-bit).

It's even possible that we are losing some quality over this since it possibly means the wrong values are being calculated.
I'll try fix it Friday when I finally have time.

**Comment by A5362** [comment]
I also have this issue with --psy-rd4, I was using 2.3.0-A

**Comment by juliobbv-p** [comment]
I've landed a few psy-rd fixes in `testing`. Can someone try encoding their problematic clips and see if the fixes solve this issue?

**Comment by BlueSwordM** [comment]
@A5362 @igorbaryshev @kuzon3 

We've finally fixed all of the (known) bugs in svt-av1-psy!

I've tested it on the friends clip Igor shared, and it works perfectly. No artifacts no matter the preset or psy-rd level.
It also **looks** better, performs more consistently, and improves encoding performance overall.

Test it with the new #testing branch and tell us how you Windows folks find it :)

**Comment by BlueSwordM** [comment]
Some Windows builds have been added to the community thread:
https://github.com/psy-ex/svt-av1-psy/discussions/128

**Comment by igorbaryshev** [comment]
@BlueSwordM nice, I indeed see no artifacts now!
Closing the issue then!
